### PR TITLE
Feature/tx handling

### PR DESCRIPTION
**Pull request by jordaniza**
Adds a few things:

General Usability:
- Better handling of decimals that otherwise round down, leading to confusing UI
- Better UI indication of when Auxo tokens are being used, versus underlying
- Conversion of Auxo tokens in the UI using the on-chain Exchange rate
- Add node.env check to automatically change certain parameters (merkle auth file, throttling limits)

Bug Fixes:
- Prevent <1 Token deposits that can cause the UI to allow deposits, even if the "max" amount has been deposited (another rounding issue)
- Fixed issue whereby trying to convert an undefined exchange rate to string causes the app to crash

Transaction improvements:
- UI updates on TX receipt and updates the state optmistically, without waiting for the onChain data

Metamask and Throttling
- Update state on every X blocks due to rate limits esp with metamask.

### CODE REVIEW

File: apps/auxo-app/src/utils/index.ts [hunk 1]
```diff
@@ -13,10 +14,19 @@ export const prettyNumber = (n?: number): string =>
...
+  decimals: number,
+  precision = 0
+): number => {
+  if (!n) return 0;
+  if (typeof n === "number") return n;
+
+  // ethers BN implementaion ignores decimal point conversion when dividing, so we need
+  // to use an alternative approach for decimals >= 0.5
+  const numerator = Big(n.toString());
+  const raw = numerator.div(Big(10).pow(decimals)).toFixed(precision);
```
Comment (Alexintosh) [inline apps/auxo-app/src/utils/index.ts:28]: Looks like this is the only place where we use `big.js`, is it worth adding the extra dependency?
You could try something like this instead:

```
export function getNormalizedNumber(number, decimals) {
  return new BigNumber(number).dividedBy(new BigNumber(10).pow(decimals));
}

export const roundDownLabel = (numberString) => {
  const number = parseFloat(numberString);
  return (Math.floor(number * 100) / 100).toString();
};
let label = roundDownLabel(getNormalizedNumber(BigNumber(balance), decimal).toString())
```

File: apps/auxo-app/src/hooks/useTransactionHandler.ts [hunk 2]
```diff
@@ -1,9 +1,55 @@
...
+  onSuccess,
+  library,
+  dispatch,
+}: {
+  tx: ContractTransaction;
+  onSuccess: CallableFunction;
+  library: ethers.providers.Web3Provider | ethers.providers.JsonRpcProvider;
+  dispatch: Dispatch<AnyAction>;
+}) => {
+  const receipt = await library.getTransactionReceipt(tx.hash);
```
Comment (Alexintosh) [inline apps/auxo-app/src/hooks/useTransactionHandler.ts:35]: No error handling on this promise.

File: apps/auxo-app/src/hooks/useTransactionHandler.ts [hunk 3]
```diff
@@ -1,9 +1,55 @@
...
+ * Awaits the transaction confirmation to see if an event is fired
+ * @param tx the transaction
+ * @param eventName string name of the event
+ * @returns whether or not the event was fired
+ */
+export const checkForEvent = async (
+  tx: ContractTransaction,
+  eventName: string
+): Promise<boolean> => {
+  const awaitedTx = await tx.wait();
```
Comment (Alexintosh) [inline apps/auxo-app/src/hooks/useTransactionHandler.ts:17]: No error handling on this promise.

File: apps/auxo-app/src/hooks/useOnChainData.ts [hunk 4]
```diff
@@ -168,20 +180,50 @@ const hasStateChanged = (old: Vault[], change: Vault[]): boolean => {
...
-      active &&
+  const shouldUpdate = useMemo(() => {
+    // do not update state if vital data missing
+    if (!chainId || !block || !block.number) return false;
+
+    // Ensure we always fetch data when the user loads the page for the first time
+    if (firstLoad) return true;
+
+    // repeated state updates can cause rpc throttling with lower block times
+    const blockFrequencyConditionMet = block.number % refreshFrequency === 0;
```
Comment (Alexintosh) [inline apps/auxo-app/src/hooks/useOnChainData.ts:198]: That's goof stuff!

File: apps/auxo-app/src/hooks/useMaxDeposit.ts [hunk 5]
```diff
@@ -2,25 +2,45 @@ import { BigNumber } from "ethers";
...
+  const decimals = useDecimals();
+  const sharesPending = vault?.userBalances?.batchBurn.shares ?? zeroBalance();
+  const exchangeRate = vault?.stats?.exchangeRate;
+
+  const amountPendingUnderlying = BigNumber.from(sharesPending.value)
+    .mul(exchangeRate?.value ?? 0)
+    .div(BigNumber.from(10).pow(decimals));
+
+  const amountPendingUnderlyingLabel = amountPendingUnderlying.div(
+    BigNumber.from(10).pow(decimals)
```
Comment (Alexintosh) [inline apps/auxo-app/src/hooks/useMaxDeposit.ts:22]: This line seems to be repeating a lot, I suggest generalizing it with something like.
```
export function getNormalizedNumber(number, decimals) {
  return new BigNumber(number).dividedBy(new BigNumber(10).pow(decimals));
}
```

File: apps/auxo-app/src/components/Vault/Actions/Withdraw/WithdrawInput.tsx [hunk 6]
```diff
@@ -34,39 +37,43 @@ function ApproveWithdrawButton({ withdraw, setWithdraw }: { withdraw: Balance, s
...
-              );
-        } finally {
-            setApproving(false)
+    const newEnterBatchBurn = async () => {
+        if (auxoContract) {
+            setApproving(true);
+            const tx = await auxoContract?.enterBatchBurn(withdraw.value)
+            await handleTx({ tx, dispatch, library, onSuccess: () => {
+                if (vault && vault.userBalances && vault.userBalances.batchBurn) {
+                    const newVault: Vault = {
```
Comment (Alexintosh) [inline apps/auxo-app/src/components/Vault/Actions/Withdraw/WithdrawInput.tsx:46]: Premise: I'm a boomer react dev but It used to be pretty unusual to update the state like this at the component level back to the redux days. The reason is that violates the separation of responsibilities by mixing state management with components in a way that is undesirable. Components are for interaction not state manipulation per se.

File: apps/auxo-app/src/components/Vault/Actions/Withdraw/WithdrawButton.tsx [hunk 7]
```diff
@@ -34,29 +38,34 @@ function WithdrawButton ({ showAvailable }: { showAvailable?: boolean }) {
...
-          setWithdrawing(false)
-        }    
-    }
+    const newMakeWithdraw = async () => {
+        if (auxoContract) {
+            setWithdrawing(true);
+            const tx = await auxoContract?.exitBatchBurn()
+            await handleTx({ tx, dispatch, library, onSuccess: () => {
+                if (vault && vault.userBalances && vault.userBalances.batchBurn) {
+                    const newVault: Vault = {
```
Comment (Alexintosh) [inline apps/auxo-app/src/components/Vault/Actions/Withdraw/WithdrawButton.tsx:47]: Same comment on state management as before

File: apps/auxo-app/src/utils/index.ts (hunk 1 above)
Comment (jordaniza) [reply apps/auxo-app/src/utils/index.ts:28]: Just need to ensure label doesn't get erroneously rounded down. Will be a solution for sure

File: apps/auxo-app/src/hooks/useTransactionHandler.ts (hunk 3 above)
Comment (jordaniza) [reply apps/auxo-app/src/hooks/useTransactionHandler.ts:17]: I think this speaks to the original point you made - the error handling is being done in the component which is far less clean than just having a single handler

File: apps/auxo-app/src/components/Vault/Actions/Withdraw/WithdrawInput.tsx (hunk 6 above)
Comment (jordaniza) [reply apps/auxo-app/src/components/Vault/Actions/Withdraw/WithdrawInput.tsx:46]: This makes a lot of sense. I think we can refactor the handle function as a reducer, then simply dispatch the action from the component.

File: apps/auxo-app/src/hooks/useTransactionHandler.ts (hunk 3 above)
Comment (dantop114) [reply apps/auxo-app/src/hooks/useTransactionHandler.ts:17]: imo error handling should be done on data sourcing and errors should bubble up to components.. in that way you can change the way components handle errors or even make other custom components

**Comment by dantop114** [comment]
good optimisation for delayed block updates retrieval!

don't know if this is feasible but (as Alex was saying) I would separate state management from components and have some kind of two-layers architecture where state management is a separate thing and components just hooks up to events fired from state changes.

File: apps/auxo-app/src/components/Vault/Actions/Withdraw/WithdrawInput.tsx (hunk 6 above)
Comment (jordaniza) [reply apps/auxo-app/src/components/Vault/Actions/Withdraw/WithdrawInput.tsx:47]: This work is now in progress: we are moving state changing to reducers and simply calling "dispatch(setThunkMakeDeposit(args))" in the button, will include in the next commit

File: apps/auxo-app/package.json [hunk 8]
```diff
@@ -23,7 +23,6 @@
     "@web3-react/portis-connector": "^6.2.9",
     "@web3-react/walletconnect-connector": "^6.2.10",
     "@web3-react/walletlink-connector": "^6.2.8",
-    "big.js": "^6.1.1",
     "buffer": "^6.0.3",
```
Comment (jordaniza) [inline apps/auxo-app/package.json:26]: @Alexintosh FYI removed external dependency, thanks for the help there

File: apps/auxo-app/src/store/vault/vault.slice.ts [hunk 9]
```diff
@@ -1,90 +1,196 @@
...
   name: 'vault',
   initialState: vaultState,
+
+  /**
+   * Extra reducers here subscribe to the async thunks 'fulfilled' hook
+   * and prepare the state transformations for the corresponding vault.
+   * To save repeating the find and update logic, we re-use the caseReducer 
+   * `setVault`, which finds and updates a single vault, according to
+   * the transformer function.
+   */
```
Comment (jordaniza) [inline apps/auxo-app/src/store/vault/vault.slice.ts:18]: Note, these are where we can define optimistic UI updates. Eventual consistency will be reached when the onChain data is loaded

**Pull request merged**